astroid/nodes/as_string.py

```diff
@@ -20,34 +20,17 @@
 # Licensed under the LGPL: https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html
 # For details: https://github.com/PyCQA/astroid/blob/master/LICENSE
 
-"""This module renders Astroid nodes as string:
-
-* :func:`to_code` function return equivalent (hopefully valid) python string
-
-* :func:`dump` function return an internal representation of nodes found
-  in the tree, useful for debugging or understanding the tree structure
-"""
-from typing import TYPE_CHECKING
-
-if TYPE_CHECKING:
-    from astroid.nodes.node_classes import (
-        Match,
-        MatchAs,
-        MatchCase,
-        MatchClass,
-        MatchMapping,
-        MatchOr,
-        MatchSequence,
-        MatchSingleton,
-        MatchStar,
-        MatchValue,
-    )
-
-# pylint: disable=unused-argument
+# Do not add typing as it will create circular import and the function cannot
+# be used with the wrong type anyway because of the visitor pattern.
```

Comment:
That's not correct. Adding types will make sure the method contents itself are checked, i.e. that all instance variables for `node` actually exist for that node type. Additionally `if TYPE_CHECKING` should not create circular imports, does it?

Reply:
In my view the function are always accessed through the visitor pattern so if ``visit_classdef``` is called, the instance will always be a ``ClassDef`` and the typing does not bring new information. Maybe we need to make the visit function private so it really cannot be accessed any other way ? Also a circular import  (according to pylint, not something that create a python error) is created by an import of Node when we also need to import the visitor in ``NodeNG.as_string``.

Reply:
Try the following with and without the argument annotation:
```py
class Custom:
    var: int

def visit(node: Custom) -> None:
    print(node.var)
    print(node.does_not_exist)
```

The point isn't that it's always `ClassDef`. That I know, it's to help check that all accessed attributes of node actually exist for the node type.

> Maybe we need to make the visit function private so it really cannot be accessed any other way ?

I don't really see a need for it.

> Also a circular import (according to pylint, not something that create a python error) is created by an import of Node when we also need to import the visitor in NodeNG.as_string.

It works fine the way it currently is. Just `visit_matchas` isn't ideal as the import is needed at runtime for the `isinstance` check.

Reply:
> Just visit_matchas isn't ideal as the import is needed at runtime for the isinstance check.

Well this MR is supposed to fix that, it's just on top of #1057 and #1058 https://github.com/PyCQA/astroid/pull/1060/commits/8df6cc4cd8c41230f0a8e516e11c55a154c79122

Also I think we're using the first answer on stackoverflow to deal with circular import right now, but the second one that follow PEP484 also make sense : https://stackoverflow.com/a/57005931/2519059, what do you think ?

Reply:
> Well this MR is supposed to fix that, it's just on top of #1057 and #1058 8df6cc4

I don't think that is worth it. Using imports has the clear advantage that you don't need test to catch errors if the class name changes. Every linter will tell you that. In addition, a string comparison prevents type checkers from doing any kind of type narrowing for the indented block.

> Also I think we're using the first answer on stackoverflow to deal with circular import right now, but the second one that follow PEP484 also make sense : https://stackoverflow.com/a/57005931/2519059, what do you think ?

If you really what to, go for it. Keep in mind however that you would need to create protocol / mixin classes for each and ever one of them. Remember also to always update them if the source changes. I guess that might be a lot of work. Using `if TYPE_CHECKING: ...` is certainly easier.